ring/partition_ring_watcher.go

```diff
@@ -81,6 +94,9 @@ func (w *PartitionRingWatcher) updatePartitionRing(desc *PartitionRingDesc) {
 	newRing := NewPartitionRing(*desc)
 
 	w.ringMx.Lock()
+	if w.delegate != nil {
+		w.delegate.OnPartitionRingChanged(&w.ring.desc, desc)
```

Comment:
Do we need to hold the lock while this runs? Does `newRing := NewPartitionRing(*desc)` make a copy of `desc`?

Reply:
Moved the callback outside of the lock